### DISCUSSION TITLE
Cache lock refresh() for long-running operations

### DISCUSSION THREAD

**Discussion by justinas55** [laravel/framework#52778] [Ideas]
Hello,

It would be nice to have a refresh() method on Cache lock instance, that would refresh lock TTL without releasing the lock. This would allow safe long-running operations without compromising with excessive TTL.
As an example such feature is implemented and documented in Symfony: https://symfony.com/doc/current/components/lock.html#expiring-locks .

Currently the closest workaround I found is this:
```
$lock = Cache::lock('test', 30);
try {
  if(!$lock->get()) {
    throw ...
  }

  $lockStart = time();

  while(!$finished) {
    if(time() - $lockStart > 15) {
       $lock->release();
       // non-atomic - another operation may acquire the lock here, and this instance then must stop/error out depending on specific case
       if(!$lock->get()) {
         throw ...
       }

       $lockStart = time();
    }

   // do actual processing
  }
} finally {
  $lock->release();
}
```

Having ->refresh(), same code could be simplified and made reliable:
```
$lock = Cache::lock('test', 30);
try {
  if(!$lock->get()) {
    throw ...
  }

  while(!$finished) {
    $lock->refresh();

   // do actual processing
  }
} finally {
  $lock->release();
}
```

Haven't found related issues, curious if this has been considered, is there known limitations to implementing this? Should be possible with Redis, probably utilizing transactions, but not sure about the other stores.

Thanks for checking!

**Comment by ringform** [comment]
I agree. In the case of long-running operations that take an unpredictable amount of time to finish (e.g. making third-party API requests), the lack of a `refresh()` function seems to defeat the purpose of acquiring cache locks to prevent race conditions in the first place.

If the cache lock TTL can't be changed and the operation exceeds that time, race conditions can occur. Therefore, for any non-trivial, long-running operation, using the current cache lock implementation is risky.

Further, the language in the [Atomic Locks documentation](https://laravel.com/docs/12.x/cache#atomic-locks) is misleading in this case:

> Atomic locks allow for the manipulation of distributed locks **without worrying about race conditions**.

**Reply by ringform** [reply]
I wrote a quick and dirty refresh lock function for Redis that works around the current limitations of the framework. It relies on closure binding to get some protected property values but I don't think the names of those properties are likely to change any time soon. And even if they do, errors are logged and the function can be updated at that point. The LUA script is used to prevent race conditions between when the cache lock key is read and when its expiration time is updated.

```php
use Illuminate\Contracts\Cache\Lock;
use Illuminate\Support\Facades\Log;

function refreshRedisLock(Lock $lock): bool
{
    $extract = \Closure::bind(function () {
        return [
            'redis'   => $this->redis   ?? null,
            'name'    => $this->name    ?? null,
            'owner'   => $this->owner   ?? null,
            'seconds' => $this->seconds ?? null,
        ];
    }, $lock, get_class($lock));

    $data = $extract();

    foreach ($data as $key => $value) {
        if ($value === null) {
            Log::error("Unable to refresh Redis lock - {$key} is missing.");
            return false;
        }
    }

    $client = $data['redis']->client();

    $script = <<<'LUA'
        local name    = tostring(KEYS[1])
        local owner   = tostring(ARGV[1])
        local seconds = tonumber(ARGV[2])

        if redis.call("GET", name) == owner then
            return redis.call("EXPIRE", name, seconds)
        end

        return 0
    LUA;

    return (bool) $client->eval(
        $script,
        [$data['name'], $data['owner'], $data['seconds']],
        1
    );
}
```

Use like so:

```php
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Redis;

$lock = Cache::lock('foo', 60);

if ($lock->get()) {
    sleep(3);

    $value = Redis::connection('default')->ttl('foo');
    Log::info('Old TTL: ' . $value);

    $refresh = refreshRedisLock($lock);
    Log::info($refresh ? 'Lock refreshed' : "Couldn't refresh lock");

    $value = Redis::connection('default')->ttl('foo');
    Log::info('New TTL: ' . $value);
}
```

Note: the function must be used inside `if ($lock->get()) { ... }` (as intended by the framework), otherwise it won't be able to read the cache lock key since it isn't created yet.

For debugging purposes, I found this command to be helpful since it allows you to see the order in which the framework, the refreshRedisLock function, and their associated LUA scripts make updates:

```bash
redis-cli -a your_password MONITOR | grep name_of_cache_lock_key
```

Edit: switched from reflection to closure binding for shorter code thanks to tip from @marius-ciclistu.

**Reply by marius-ciclistu** [reply]
Closure binding is shorter as code and has the same outcone.

**Reply by ringform** [reply]
@marius-ciclistu Good tip; I updated my code. Hadn't heard of closure binding ([1](https://www.php.net/manual/en/closure.bind.php), [2](https://php-dictionary.readthedocs.io/en/latest/dictionary/closure-binding.ini.html), [3](https://markbakeruk.net/2021/10/24/closure-binding-for-unit-testing-of-private-and-protected-methods/)) before.

**Comment by macropay-solutions** [comment]
Memcache does not allow extending TTL without resetting the key. So what you did in your code example is what symfony is doing behind the scenes.

**Reply by macropay-solutions** [reply]
Sorry. It does a replace not a release and get.

**Reply by macropay-solutions** [reply]
For memcache could look like this:
```php
    public function refresh(?int $seconds = null): bool
    {
        return $this->isOwnedByCurrentProcess() && $this->memcached->replace(
            $this->name,
            $this->owner,
            $seconds ?? $this->seconds
        );
    }

```

**Reply by macropay-solutions** [reply]
@justinas55 @ringform 
What do you think about this backward compatible [POC](https://github.com/macropay-solutions/maravel-framework/pull/31/files)  ?

**Reply by ringform** [reply]
@macropay-solutions Looks good! It's much nicer to have an official refresh function than a hacky workaround. Are you planning on making a PR for the _Laravel_ framework too?

**Reply by marius-ciclistu** [reply]
@ringform no. Feel free to do it if you want. They will not accept it without changing the interface/contract, that means v13 should be the target.

**Reply by bytestream** [reply]
Check https://github.com/laravel/framework/pull/58349

**Reply by marius-ciclistu** [reply]
@bytestream did you had to change something for v12 vs the above poc?

**Reply by bytestream** [reply]
@marius-ciclistu I didn't modify the interface, and some of the code is slightly different to your poc.

**Reply by marius-ciclistu** [reply]
You did not add refresh to UniqueLock. You missed that or it was intentional?

Interesting the dynamo db owner condition. I did it like for the incrementing.
